### PR TITLE
load the next level of navtree databases

### PR DESCRIPTION
**Pull request by jchris**
I need help with how to do some immutable stuffs. notes inline.

### CODE REVIEW

File: src/new-dash/schema/schema-tree.js [hunk 1]
```diff
@@ -41,6 +41,17 @@ const queryForSchema = (client, dbPath) => {
...
       databases: getInstancesOf("databases")
+    }).then((resp)=>{
+      console.log("dataenbases", dbPath, resp)
+      return Promise.all(resp.databases.data.map((db, i)=>{
+        var searchPath = dbPath.toJS().concat([db.name])
+        console.log("dataenbasesx", searchPath)
+        client.queryWithPrivilegesOrElse(searchPath, KeyType.ADMIN, {},
+        {databases:  getInstancesOf("databases")}).then((resp2) => {
+          console.log("dataenbases2", searchPath, resp2.databases.data)
+          resp.databases.data[i].databases = resp2.databases
```
Comment (jchris) [inline src/new-dash/schema/schema-tree.js]: I think if this response gets incorporated into the data structure correctly we can use it later when we unfold, and only have to fetch the unfolded children's children. @erickpintor

File: src/new-dash/schema/schema-tree.js [hunk 2]
```diff
@@ -59,17 +79,7 @@ export const loadSchemaTree = (client, dbPath = []) => (dispatch, getState) => {
   const dbNode = getState().get("schema").getIn(nodePath, Map())
   const info = dbNode.get("info", Map.of("name", nodePath.last() || "/"))
 
-  if (dbNode.get("loaded", false))
-    return Promise.resolve()
-
```
Comment (jchris) [inline src/new-dash/schema/schema-tree.js]: I removed this check... I'm not sure what it means or who else is setting it / depending on it.

**Comment by jchris** [comment]
I think it's ready to merge

File: src/new-dash/schema/components/nav-db-tree.js [hunk 3]
```diff
@@ -12,6 +13,7 @@ class NavDBTree extends Component {
   }
 
   componentDidMount() {
+    this.props.client && this.props.dispatch(loadSchemaTree(this.props.client, []))
```
Comment (erickpintor) [inline src/new-dash/schema/components/nav-db-tree.js]: I don't think this should be triggered by the UI. I feel like if we want the first level of the schema tree to be always present, we should encode and test that logic into the schema tree itself.

File: src/new-dash/schema/components/nav-db-tree.js [hunk 4]
```diff
@@ -38,7 +40,7 @@ class NavDBTree extends Component {
         key,
         url: key,
         links: this.databaseLinks(db, link.links || []),
-        isExpanded: (typeof link.isExpanded === "undefined" ? true : link.isExpanded)
+        isExpanded: (typeof link.isExpanded === "undefined" ? false : link.isExpanded)
```
Comment (erickpintor) [inline src/new-dash/schema/components/nav-db-tree.js]: We'll probably want to get a little fancier here. If we just default to collapsed, the tree for the selected database will be closed when the user refreshes the page.

File: src/new-dash/schema/components/nav-db-tree.js [hunk 5]
```diff
@@ -48,6 +50,11 @@ class NavDBTree extends Component {
       .toJS()
   }
 
+  onExpandClick(link) {
+    var path = link.key.split('/').slice(1);
```
Comment (erickpintor) [inline src/new-dash/schema/components/nav-db-tree.js]: This works but, ideally we want to change the `databaseTree` function at `schema/derived-data` to also contain the database path for each level of the database tree. That way, we keep this logic contained within the `schema` module itself.

File: src/new-dash/schema/components/nav-db-tree.js [hunk 6]
```diff
@@ -48,6 +50,11 @@ class NavDBTree extends Component {
...
+    this.props.client && this.props.dispatch(loadSchemaTree(this.props.client, path))
```
Comment (erickpintor) [inline src/new-dash/schema/components/nav-db-tree.js]: You can trust the client will be present at this point. The main container won't render client dependent resources unless it's present.

File: src/new-dash/schema/schema-tree.js [hunk 7]
```diff
@@ -32,15 +32,35 @@ const toDatabase = (info, loaded = false, result = {}) => {
...
   ref => q.Get(ref)
 )
 
-const queryForSchema = (client, dbPath) => {
+const queryForSchema = (dispatch, client, dbPath) => {
   return Promise.all([
     client.queryWithPrivilegesOrElse(dbPath, KeyType.ADMIN, {}, {
       databases: getInstancesOf("databases")
+    }).then((resp)=>{
+      Promise.all(resp.databases.data.map((db, i)=>{
```
Comment (erickpintor) [inline src/new-dash/schema/schema-tree.js]: I'm sure this works fine but, I feel like we're braking the single responsibility principle a little. The `queryForSchema` both queries and updates the schema tree. It also returns as well as it has side effects.

We should at least have a better name or, if possible, take this out and put it as a continuation for the first query at `loadSchemaTree` function.

File: src/new-dash/schema/schema-tree.js (hunk 2 above)
Comment (erickpintor) [inline src/new-dash/schema/schema-tree.js]: We need this so we don't keep calling the server as the user navigates through the schema tree when the data is already loaded.

I imagine in the future we'll have a timestamp of a asynchronous process that check for changes on the schema and updates the UI. For now, this check provides performance and less network traffic when users are exploring their nested databases.

**Comment by erickpintor** [comment]
Implemented at: https://github.com/fauna/dashboard/pull/170

**Pull request closed without merging**